cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -362,117 +360,78 @@ fn add_duplicate_occurrence(
 
 /// The actual parser is intended to be entirely opaque from outside users.
 impl YaccParser<'_> {
-    pub(crate) fn new(header: &mut Header, src: String) -> YaccParser {
+    pub(crate) fn new(yacc_kind: YaccKind, src: &str) -> YaccParser {
         YaccParser {
-            header,
-            yacc_kind: None,
+            yacc_kind,
             src,
             num_newlines: 0,
             ast: GrammarAST::new(),
             global_actiontype: None,
         }
     }
 
+    #[cfg(test)]
+    pub(crate) fn from_src(src: &str) -> Result<YaccParser, Vec<HeaderError>> {
+        use crate::header::HeaderErrorKind;
+        use crate::span::Location;
+        let (header, pos) = GrmtoolsSectionParser::new(src, true).parse()?;
+        let yacc_kind = header
+            .get("yacckind")
+            .map(|(_, val)| YaccKind::try_from(val))
+            .transpose()
+            .map_err(|e| vec![e])?;
+        if let Some(yacc_kind) = yacc_kind {
+            Ok(YaccParser {
+                yacc_kind,
+                src,
+                num_newlines: 0,
+                ast: GrammarAST::new(),
+                global_actiontype: None,
+            })
+        } else {
+            Err(vec![HeaderError {
+                kind: HeaderErrorKind::InvalidEntry("yacckind"),
+                locations: vec![Location::Span(Span::new(pos, pos))],
+            }])
+        }
+    }
+
     pub(crate) fn parse(&mut self) -> YaccGrammarResult<usize> {
         let mut errs = Vec::new();
+        let (_, pos) = GrmtoolsSectionParser::new(self.src, false).parse().unwrap();
```

Comment:
I wasn't expecting `unwrap` here -- should it be `?`?

Reply:
Ahh, good catch.  That was supposed to be a temporary fix until I migrated to `Header<Span>` so we could actually return ` YaccGrammarError` at this point.  I.e. this is unwrap because the error still has a `Location`, even though it is always a `Location::Span(...)`.

Reply:
So fixing this was a little more involved than I like (Which is why I was so happy to have avoided doing so, until you found I had forgotten this unwrap!)

But should be fixed in [bb6225f](https://github.com/softdevteam/grmtools/pull/554/commits/bb6225f5e6dc0af2e200230f4bf4000e76864f44)

Reply:
Wow, yeah, that's a big change!

Reply:
Woops, I actually forgot the specific `unwrap()` here, since that unwrap was hidding the type conversion error.

Should really be fixed in [3df2592](https://github.com/softdevteam/grmtools/pull/554/commits/3df259282deb755e4e66fe785eb2ff9e6eefb1c8)